src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala

```diff
@@ -28,14 +48,54 @@ private[next] final class NextIterableOnceOpsExtensions[A, CC[_], C](
    *
    * @note This will force the evaluation of the Iterator.
    */
-  def groupMapReduce[K, B](key: A => K)(f: A => B)(reduce: (B, B) => B): immutable.Map[K, B] = {
-    val m = mutable.Map.empty[K, B]
-    col.foreach { elem =>
-      m.updateWith(key = key(elem)) {
-        case Some(b) => Some(reduce(b, f(elem)))
-        case None    => Some(f(elem))
+  def groupMapReduce[K, V](key: A => K)(f: A => V)(reduce: (V, V) => V): immutable.Map[K, V] =
+    groupMapGenGen(key)(f).reduceValues(reduce)
+}
+
+private[next] object NextIterableOnceOpsExtensions {
+  final class GroupMapGenGen[A, K, V] private[NextIterableOnceOpsExtensions](
+    col: IterableOnceOps[A, AnyConstr, _],
+    key: A => K,
+    f: A => V
+  ) {
+    def reduceValues(reduce: (V, V) => V): immutable.Map[K, V] =
+      reduceValuesAs(immutable.Map)(reduce)
+
+    def reduceValuesAs[MC](resultFactory: Factory[(K, V), MC])(reduce: (V, V) => V): MC = {
```

Comment:
I suggest replacing “as” with “to”:

```suggestion
    def reduceValuesTo[MC](resultFactory: Factory[(K, V), MC])(reduce: (V, V) => V): MC = {
```

This would be more consistent with the existing `to` method that collects the elements of a collection into another collection type.

~~~ scala
map.viewGroupsBy(f).reduceValuesTo(TreeMap)(_  + _)
~~~